src/MyInfoGovClient.class.ts

```diff
@@ -1,609 +1,331 @@
+   * consent to provide
+   * @param config.relayState Optional state to be forwarded to the redirect endpoint
+   * via query parameters
+   * @param config.singpassEserviceId Optional alternative e-service ID.
+   * Defaults to the e-serviceId provided in the constructor.
+   * @param config.redirectEndpoint Optional alternative redirect endpoint.
+   * Defaults to the endpoint provided in the constructor.
+   * @returns The URL which the user should visit to log in to SingPass
+   * and consent to providing the given attributes.
    */
-  getPersonBasic ({
-    uinFin,
+  createRedirectURL({
+    purpose,
     requestedAttributes,
-    txnNo,
-    singpassEserviceId: seId,
-  }: IPersonBasicRequest): Promise<IPersonBasic> {
-    if (!requestedAttributes || requestedAttributes.length === 0) {
-      requestedAttributes = ALL_ATTRIBUTES
+    relayState,
+    singpassEserviceId,
+    redirectEndpoint,
+  }: IAuthRequest): string {
+    const queryParams = {
+      purpose,
+      attributes: requestedAttributes.join(),
+      state: relayState ?? '',
+      client_id: this.clientId,
+      redirect_uri: redirectEndpoint ?? this.redirectEndpoint,
+      sp_esvcId: singpassEserviceId ?? this.singpassEserviceId,
     }
+    return `${this.baseAPIUrl}${Endpoint.Authorise}?${qs.stringify(
+      queryParams,
+    )}`
+  }
 
-    const url = this.baseUrl + path.join(ENDPOINT.personBasic, uinFin) + '/'
-    const nonce = crypto.randomBytes(32).toString('base64')
-    const timestamp = Date.now()
-
-    const singpassEserviceId = seId || this.singpassEserviceId
-
-    // Construct the request basestring
-    const basestring = this._formulateBaseString({
-      httpMethod: 'GET',
-      url,
-      appId: this.appId,
-      clientId: this.clientId,
-      singpassEserviceId,
-      nonce,
-      requestedAttributes,
-      timestamp,
-      txnNo,
-    })
-
-    // Generate the request signature by signing the constructed basestring
-    // with private key
-    const signature = this._signBaseString(
-      basestring,
-      this.privateKey,
-      'base64',
-    )
+  /**
+   * Retrieves the given MyInfo attributes from the Person endpoint after
+   * the client has logged in to SingPass and consented to providing the given
+   * attributes.
+   * @param authCode Authorisation code given by MyInfo
+   * @param requestedAttributes Attributes to request from Myinfo. Should correspond
+   * to the attributes provided when initiating SingPass login.
+   * @returns Object containing access token used to retrieve the data,
+   * the NRIC/FIN and the data
+   */
+  async getPerson(
+    authCode: string,
+    requestedAttributes: string[],
+  ): Promise<IPersonResponse> {
+    // Obtain access token
+    let accessToken: string
+    try {
+      accessToken = await this._getAccessToken(authCode)
+    } catch (err) {
+      throw new Error(
+        `The following error occurred while retrieving the access token: ${err}`,
+      )
+    }
```

Comment:
Merging the `_getAccessToken` into `getPerson` makes sense only if we're absolutely certain that MyInfo will never change their access token policies to be one-use only.

It may appear convenient now, but an SDK shouldn't assume responsibility for how the access token is to be handled. It would be less presumptive if `_getAccessToken` is sharded out into a public function to be called explicitly by users of the SDK.

Token exchange is also an explicit step in the OAuth flow that should not be obscured from users, and it would also save future work if e.g. MyInfo changes their policy to allow access token reuse.

Reply:
Is there a use case for retrieving the access token but subsequently _not_ calling the Person endpoint? I wrote it this way because I thought those two actions are always done together. Since we return the access token, the user is free to do what they want with it.

```
// current implementation
const { accessToken, data } = await client.getPerson(authCode, requestedAttributes)
doSomethingWith(accessToken)

// versus if we separate the methods
const accessToken = await client.getAccessToken(authCode)
const data = await client.getPerson(accessToken, requestedAttributes)
doSomethingWith(accessToken)
```

In future, if access tokens can be reused, we can just expose a `getPersonFromAccessToken` method. We can do so now as well, but there isn't really a need for it, and it might be misleading to any potential users who aren't aware that access tokens currently cannot be reused. Plus, if MyInfo's implementation of refreshing is different (e.g. if they return a refresh token with the original call to the /token endpoint, and require the refresh token to create a new access token), then we'd need to deprecate `getPersonFromAccessToken` without ever having a use for it.

Reply:
> Is there a use case for retrieving the access token but subsequently not calling the Person endpoint?

Come to think of it, there is. The successful exchange of auth code for an access token is proof of authentication, and we can already parse the UIN/FIN. Given that the SingPass SAML flow is aging, if we needed it, we could migrate our Singpass login to use this step.

> misleading to any potential users who aren't aware that access tokens currently cannot be reused

I don't think user misunderstanding is a real consideration. They could also misconstrue the auth code as an access token and try to call the function twice.

> we may need to make breaking changes to getPersonFromAccessToken

In this case, I would extend the API to `getPerson(accessToken[, refreshToken])`, or introduce a new `refreshAccessToken(refreshToken)` API for convenience.

I think the crux of the issue is finding the correct level of abstraction. I think the design of this SDK should reflect its job as a wrapper around the OAuth 2 protocol, in the context of the MyInfo data provider. This means we should not abstract away distinct steps in the protocol, or remove control from the user of when each step takes place, simply because our own use cases or current MyInfo policies make it pointless to do so.